### PR TITLE
emlinux3: cve-check: Add pre-download CVE database feature

### PR DESCRIPTION
**Pull request by masami256**
# Purpose of pull request

## commit:  Add pre-download CVE database feature 

This commit adds the feature to download a prepared database file to reduce access to the NVD's API server. Typically,cve_check.py creates a
database file using the NVD JSON API. However, it takes time to retrieve all CVE information via the API. Therefore, it would be beneficial to
reduce this time. Additionally, the NVD API server can sometimes be unstable for short or long periods. In such cases, downloading a
prepared database file instead of creating it from scratch would be preferable.

To enable download pre-downloaded CVE database feature, usage --cve-db-predownload option and set URL to CVE_DB_PREDOWNLOAD_URL variable in conf/local.conf.
The predownloaded CVE database name should be "nvd_cve_db.db".
e.g.

Add CVE_DB_PREDOWNLOAD_URL in conf/local.conf.

```
CVE_DB_PREDOWNLOAD_URL="http://127.0.0.1:8000/nvd_cve_db.db"
```

Then run cve_check.py with

```
cve_check.py --image emlinux-image-base \
--output-format text,json --cve-db-predownload
```

It download pre-downloaded CVE database when one of following contision is met.

- CVE database is not present
- CVE database is present but it is old

## commit: cve-check: Add update-cve-databese-only

Add update-cve-databese-only that only updates CVE database. It purpose is to prepare CVE database then distribute it to other  machines.

# Test
### Download pre-downloaded file

Test patterns

|no|download pre-download file|update CVE database|
|--|--|--|
|1|Yes(CVE database file is not present in build/download/CVE directory|No|
|2|Yes(CVE database file is  present in build/download/CVE directory|No|
|3|Yes(CVE database file is  present in build/download/CVE directory|Yes(Downloaded file is old. may be it is not updated by server)|
|4|No|Yes (normal use case before applying this PR)|

Prepare two CVE database one is up to dated and the other is bit old. for example prepare nvd_cve_db.db.new annd nvd_cve_db.db.old .
Copy .new or .old file to nvd cve db.db which depends on test pattern separately.
Run "python -m http.server" to host CVE database file.

Run cve-check.py for test 1 to 3.

```
 cve_check.py \
--image emlinux-image-base \
--output-format text,json \
--cve-db-predownload 
```

Run cve-check.py for test 4.

```
 cve_check.py \
--image emlinux-image-base \
--output-format text,json \
```

### Update CVE database only

Run "python -m http.server" to host CVE database file.

Run cve-check.py

```
cve_check.py \
--image emlinux-image-base \
--output-format text,json \
--cve-db-predownload \
--update-cve-databese-only
```

### Download pre-downloaded file

test pattern 1

```
build@3d4795c39605:~/work$ rm -f build/downloads/CVE/nvd_cve_db.db 
build@3d4795c39605:~/work$ cve_check.py --image emlinux-image-base --output-format text,json  --cve-db-predownload 
2024-07-26 05:37:19,749:INFO: Predownload CVE database file.
2024-07-26 05:37:19,749:INFO: Download CVE database file from http://127.0.0.1:8000/nvd_cve_db.db.
2024-07-26 05:37:19,934:INFO: Download CVE database file was succeeded.
2024-07-26 05:37:22,557:INFO: Last database update is in 1day so skip NVD database update
2024-07-26 05:37:22,558:INFO: Update debian CVE database
2024-07-26 05:37:22,558:INFO: Last database update is in 1day so skip Debian CVE database update
2024-07-26 05:37:22,558:INFO: Update cip-kernel-sec
2024-07-26 05:37:23,389:INFO: Checking CVEs ...
2024-07-26 05:38:00,154:INFO: CVE check finished. CVE check results are stored in /home/build/work/build/tmp/deploy/cve/emlinux-image-base-emlinux-bookworm-qemu-arm64
```

test pattern 2

before testing, modify CVE database.

```
build@3d4795c39605:~/work$ sqlite3 ./build/downloads/CVE/nvd_cve_db.db 
SQLite version 3.40.1 2022-12-28 14:03:47
Enter ".help" for usage hints.
sqlite> select * from meta;
1|2024-07-26T05:33:59.720785
sqlite> .schema meta
CREATE TABLE META (ID NUMBER UNIQUE, LASTMODIFIED TEXT);
sqlite> select * from meta;
1|2024-07-26T05:33:59.720785
sqlite> update meta set lastmodified="2024-07-25T01:33:59.720785" where id=1;
sqlite> select * from meta;
1|2024-07-25T01:33:59.720785
```

```
build@3d4795c39605:~/work$ cve_check.py --image emlinux-image-base --output-format text,json  --cve-db-predownload 
2024-07-26 05:44:47,682:INFO: Predownload CVE database file.
2024-07-26 05:44:47,682:INFO: Download CVE database file from http://127.0.0.1:8000/nvd_cve_db.db.
2024-07-26 05:44:47,885:INFO: Download CVE database file was succeeded.
2024-07-26 05:44:50,148:INFO: Last database update is in 1day so skip NVD database update
2024-07-26 05:44:50,148:INFO: Update debian CVE database
2024-07-26 05:44:50,148:INFO: Last database update is in 1day so skip Debian CVE database update
2024-07-26 05:44:50,148:INFO: Update cip-kernel-sec
2024-07-26 05:44:50,796:INFO: Checking CVEs ...
2024-07-26 05:45:27,497:INFO: CVE check finished. CVE check results are stored in /home/build/work/build/tmp/deploy/cve/emlinux-image-base-emlinux-bookworm-qemu-arm64
```

test pattern 3

before testing,  copy old CVE database to build/download/CVE directory and web server distributes old CVE database.

```
build@3d4795c39605:~/work$ cp web/nvd_cve_db.db.20240719 build/downloads/CVE/nvd_cve_db.db 
build@3d4795c39605:~/work$ cve_check.py --image emlinux-image-base --output-format text,json  --cve-db-predownload 
2024-07-26 05:48:36,019:INFO: Predownload CVE database file.
2024-07-26 05:48:36,019:INFO: Download CVE database file from http://127.0.0.1:8000/nvd_cve_db.db.
2024-07-26 05:48:36,227:INFO: Download CVE database file was succeeded.
2024-07-26 05:48:38,985:INFO: Update NVD CVE database
2024-07-26 05:48:41,198:INFO: Update last modified date
2024-07-26 05:48:41,758:INFO: Update debian CVE database
2024-07-26 05:48:41,758:INFO: Last database update is in 1day so skip Debian CVE database update
2024-07-26 05:48:41,759:INFO: Update cip-kernel-sec
2024-07-26 05:48:42,393:INFO: Checking CVEs ...
2024-07-26 05:49:19,536:INFO: CVE check finished. CVE check results are stored in /home/build/work/build/tmp/deploy/cve/emlinux-image-base-emlinux-bookworm-qemu-arm64
```

test pattern 4

```
build@3d4795c39605:~/work$ cp web/nvd_cve_db.db.20240719 build/downloads/CVE/nvd_cve_db.db 
build@3d4795c39605:~/work$ cve_check.py --image emlinux-image-base --output-format text,json 
2024-07-26 05:49:44,496:INFO: Update NVD CVE database
2024-07-26 05:49:47,136:INFO: Update last modified date
2024-07-26 05:49:48,016:INFO: Update debian CVE database
2024-07-26 05:49:48,016:INFO: Last database update is in 1day so skip Debian CVE database update
2024-07-26 05:49:48,016:INFO: Update cip-kernel-sec
2024-07-26 05:49:48,667:INFO: Checking CVEs ...
2024-07-26 05:50:25,641:INFO: CVE check finished. CVE check results are stored in /home/build/work/build/tmp/deploy/cve/emlinux-image-base-emlinux-bookworm-qemu-arm64
```
### Update CVE database only

```
build@5abcfbd7ac81:~/work$ cve_check.py \
--image emlinux-image-base \
--output-format text,json \
--cve-db-predownload \
--update-cve-databese-only
2024-07-24 01:20:06,942:INFO: predownload file is old, download it again.
2024-07-24 01:20:06,942:INFO: Download CVE database file from http://127.0.0.1:8000/nvd_cve_db.db.
2024-07-24 01:20:07,146:INFO: Download CVE database file was succeeded.
2024-07-24 01:20:07,845:INFO: Update NVD CVE database
2024-07-24 01:20:11,465:INFO: Update last modified date
2024-07-24 01:20:11,800:INFO: Finish CVE database update.
```

### CODE REVIEW

File: scripts/lib/python/cve/nvd_cve.py [hunk 1]
```diff
@@ -258,37 +258,86 @@ def initialize_nvd_cve_db(conn):
...
+            else:
+                raw_data = r.read()
+
+            r.close()
+        except Exception as e:
+            logger.debug(f"CVE databese download: received error ({e}), retrying")
+            time.sleep(6)
+            pass
+        else:
+            with open(db_file, "wb") as f:
```
Comment (hiraku-wfs) [inline scripts/lib/python/cve/nvd_cve.py:280]: When the basenames of the predownload_url differs from the one of the db_file, we should rename the predownloaded file.

File: scripts/lib/python/cve/nvd_cve.py [hunk 2]
```diff
@@ -258,37 +258,86 @@ def initialize_nvd_cve_db(conn):
...
+    logger.debug(f"Initialize nvd cve database {db_file}")
+    initialize_nvd_cve_db(conn)
+
+    skip_db_update, last_modified = check_skip_db_update(conn)
+
+    if not skip_db_update and predownload_url:
+        # predownload database file is old, download latest file
+        conn.close()
+        logger.info("predownload file is old, download it again.")
+        if not download_nvdb(predownload_url, db_file):
```
Comment (hiraku-wfs) [inline scripts/lib/python/cve/nvd_cve.py]: If the db_file was downloaded right before (L313) and its last_modified is too old, then the same old file will be meaninglessly downloaded again.

File: scripts/lib/python/cve/nvd_cve.py [hunk 3]
```diff
@@ -258,37 +258,86 @@ def initialize_nvd_cve_db(conn):
...
+            return None
+        conn = sqlite3.connect(db_file)
+        # re-check last modified date
+        skip_db_update, last_modified = check_skip_db_update(conn)
+
     if not skip_db_update:
+        logger.info("Update NVD CVE database")
         if not fetch_all_cves(db_file, conn, last_modified, nvd_api_key):
             return None
```
Comment (hiraku-wfs) [inline scripts/lib/python/cve/nvd_cve.py]: In general, we should not return while a connection is open.

File: scripts/lib/python/cve/nvd_cve.py (hunk 1 above)
Comment (masami256) [reply scripts/lib/python/cve/nvd_cve.py:280]: It doesn't use file name from user given URL when saving received data.

```
build@ab0904b04bbd:~/work$ ls build/downloads/CVE
ls: cannot access 'build/downloads/CVE': No such file or directory
build@ab0904b04bbd:~/work$ cve_check.py \
--image emlinux-image-base \
--output-format text,json \
--cve-db-predownload \
--update-cve-databese-only
2024-07-24 06:49:59,843:INFO: Download CVE database file from http://127.0.0.1:8000/nvd_cve_db.db.20240724.
2024-07-24 06:50:00,025:INFO: Download CVE database file was succeeded.
2024-07-24 06:50:00,034:INFO: Last database update is in 1day so skip NVD database update
2024-07-24 06:50:00,034:INFO: Finish CVE database update.
build@ab0904b04bbd:~/work$ ls build/downloads/CVE/
nvd_cve_db.db
```

File: scripts/lib/python/cve/nvd_cve.py (hunk 3 above)
Comment (masami256) [reply scripts/lib/python/cve/nvd_cve.py]: thanks. I'll fix it.

File: scripts/lib/python/cve/nvd_cve.py (hunk 2 above)
Comment (masami256) [reply scripts/lib/python/cve/nvd_cve.py]: I see.

File: scripts/lib/python/cve/nvd_cve.py (hunk 1 above)
Comment (hiraku-wfs) [reply scripts/lib/python/cve/nvd_cve.py:280]: Oh, I misunderstood. Data is saved to the db_file via raw_data, so it is okay.

File: scripts/lib/python/cve/nvd_cve.py (hunk 2 above)
Comment (masami256) [reply scripts/lib/python/cve/nvd_cve.py]: Fixed.

File: scripts/lib/python/cve/nvd_cve.py (hunk 3 above)
Comment (masami256) [reply scripts/lib/python/cve/nvd_cve.py]: Fixed.

**Comment by masami256** [comment]
@hiraku-wfs I pushed new commit. PTAL

File: scripts/lib/python/cve/nvd_cve.py [hunk 4]
```diff
@@ -258,38 +257,83 @@ def initialize_nvd_cve_db(conn):
 
         c.close()
 
-def update_nvd_db(dl_dir, nvd_api_key):
-    db_file = f"{dl_dir}/{CVE_DATABASE_NAME}"
+def fetch_predownloaded_db(predownload_url, db_file):
```
Comment (hiraku-wfs) [inline scripts/lib/python/cve/nvd_cve.py]: "Predownload" is the role of this function, so this should be `predownload_db()`.

File: scripts/lib/python/cve/nvd_cve.py [hunk 5]
```diff
@@ -258,38 +257,83 @@ def initialize_nvd_cve_db(conn):
...
+    conn = sqlite3.connect(db_file)
+    logger.debug(f"Initialize nvd cve database {db_file}")
+    initialize_nvd_cve_db(conn)
+
+    skip_db_update, last_modified = check_skip_db_update(conn)
+
+    if not skip_db_update and predownload_url:
+        # predownload database file is old, download latest file
+        conn.close()
+        logger.info("Fetch predownload CVE database file.")
```
Comment (hiraku-wfs) [inline scripts/lib/python/cve/nvd_cve.py]: This message should be "Predownload CVE database file."

**Comment by masami256** [comment]
@hiraku-wfs I pushed new commit. PTAL. Also, I update PR comment .

**Comment by hiraku-wfs** [comment]
Thanks!

**Pull request merged**